File: app/src/main/java/ru/strcss/projects/moneycalc/moneycalcandroid/utils/logic/ComponentsUtils.java
package ru.strcss.projects.moneycalc.moneycalcandroid.utils.logic;

import java.io.IOException;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import retrofit2.HttpException;
import ru.strcss.projects.moneycalc.moneycalcdto.entities.FinanceSummaryBySection;
import ru.strcss.projects.moneycalc.moneycalcdto.entities.SpendingSection;

public class ComponentsUtils {
    private static final String messageRegex = "\"message\":\"(.*?)\"";
    private static final Pattern messageGetterPattern = Pattern.compile(messageRegex);

    public static SpendingSection getSpendingSectionByInnerId(List<SpendingSection> spendingSections, int id) {
        if (spendingSections == null)
            return null;
        for (SpendingSection spendingSection : spendingSections) {
            if (spendingSection.getSectionId() == id)
                return spendingSection;
        }
        return null;
    }

    public static int getPositionBySpendingSectionInnerId(List<SpendingSection> spendingSections, int sectionInnerId) {
        if (spendingSections != null) {
            for (int i = 0; i < spendingSections.size(); i++) {
                if (spendingSections.get(i).getSectionId() == sectionInnerId)
                    return i;
            }
        }
        return 0;
    }

    public static Integer getSpendingSectionInnerIdByPosition(List<SpendingSection> spendingSections, int position) {
        if (spendingSections != null) {
            if (spendingSections.size() > position)
                return spendingSections.get(position).getSectionId();
        }
        return null;
    }

    public static Integer getLogoIdBySectionId(List<SpendingSection> spendingSections, Integer sectionId) {
        SpendingSection spendingSection = getSpendingSectionByInnerId(spendingSections, sectionId);
        return spendingSection == null ? null : spendingSection.getLogoId();
    }

    public static FinanceSummaryBySection getFinanceSummaryBySectionById(List<FinanceSummaryBySection> financeList, int id) {
        for (FinanceSummaryBySection financeSummaryBySection : financeList) {
            if (financeSummaryBySection.getSectionId() == id)
                return financeSummaryBySection;
        }
        return null;
    }

    public static String getErrorBodyMessage(Throwable t) {
        if (t instanceof HttpException)
            return getErrorBodyMessage((HttpException) t);
        else
            return t.getMessage();
    }

    public static String getErrorBodyMessage(HttpException ex) {
        try {
            String errorJSON = ex.response().errorBody().string();
            System.err.println("errorJSON = " + errorJSON);
            final Matcher messageMatcher = messageGetterPattern.matcher(errorJSON);

            if (messageMatcher.find()) {
                return messageMatcher.group(1);
            } else {
                return errorJSON;
            }
        } catch (IOException e1) {
            e1.printStackTrace();
            return ex.getMessage();
        }
    }
}

File: app/src/main/java/ru/strcss/projects/moneycalc/moneycalcandroid/utils/logic/ComponentsUtils.kt
package ru.strcss.projects.moneycalc.moneycalcandroid.utils.logic

import retrofit2.HttpException
import ru.strcss.projects.moneycalc.moneycalcdto.entities.SpendingSection
import ru.strcss.projects.moneycalc.moneycalcdto.entities.statistics.SummaryBySection
import java.io.IOException
import java.util.regex.Matcher
import java.util.regex.Pattern

class ComponentsUtils {

    companion object {
        private const val messageRegex = "\"userMessage\":\"(.*?)\""
        private const val messageHtmlRegex = "<title>(.*?)</title>"
        private val messageGetterPattern = Pattern.compile(messageRegex)
        private val messageHtmlGetterPattern = Pattern.compile(messageHtmlRegex)

        fun getSpendingSectionByInnerId(spendingSections: List<SpendingSection>?, id: Int): SpendingSection? {
            if (spendingSections == null)
                return null
            for (spendingSection in spendingSections) {
                if (spendingSection.sectionId == id)
                    return spendingSection
            }
            return null
        }

        fun getPositionBySpendingSectionInnerId(spendingSections: List<SpendingSection>?, sectionInnerId: Int): Int {
            if (spendingSections != null) {
                for (i in spendingSections.indices) {
                    if (spendingSections[i].sectionId == sectionInnerId)
                        return i
                }
            }
            return 0
        }

        fun getSpendingSectionInnerIdByPosition(spendingSections: List<SpendingSection>?, position: Int): Int? {
            if (spendingSections != null) {
                if (spendingSections.size > position)
                    return spendingSections[position].sectionId
            }
            return null
        }

        fun getLogoIdBySectionId(spendingSections: List<SpendingSection>?, sectionId: Int?): Int? {
            val spendingSection = getSpendingSectionByInnerId(spendingSections, sectionId!!)
            return spendingSection?.logoId
        }

        fun getLogoIdByName(spendingSections: List<SpendingSection>?, name: String?): Int? {
            return spendingSections
                    ?.filter { spendingSection -> spendingSection.name == name }
                    ?.map { spendingSection -> spendingSection.logoId }
                    ?.firstOrNull()
        }

        fun getStatsSummaryBySectionById(financeList: List<SummaryBySection>, id: Int): SummaryBySection? {
            for (financeSummaryBySection in financeList) {
                if (financeSummaryBySection.sectionId == id)
                    return financeSummaryBySection
            }
            return null
        }

        fun getErrorBodyMessage(t: Throwable): String {
            return if (t is HttpException)
                getErrorBodyMessage(t)
            else
                t.message ?: ""
        }

        fun getErrorBodyMessage(ex: HttpException): String {
            return try {
                val errorBody = ex.response().errorBody()
                val errorJSON = errorBody!!.string()

                System.err.println("errorJSON = $errorJSON")
                val messageMatcher: Matcher?

                if (errorBody.contentType()?.type().equals("text")) {
                    messageMatcher = messageHtmlGetterPattern.matcher(errorJSON)
                } else {
                    messageMatcher = messageGetterPattern.matcher(errorJSON)
                }

                if (messageMatcher.find()) {
                    messageMatcher.group(1)
                } else {
                    errorJSON
                }
            } catch (e1: IOException) {
                e1.printStackTrace()
                ex.message ?: ""
            }
        }
    }
}
